src/systems/filecoin_markets/storage_market/storage_market_actor_state.go

```diff
@@ -0,0 +1,288 @@
+
+	st._assertDealStartAfterCurrEpoch(rt, p)
+
+	// deal ends before it starts
+	if p.EndEpoch() <= p.StartEpoch() {
+		rt.Abort("sma._assertValidDealTimingAtPublish: deal ends before it starts.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertValidDealMinimum(rt Runtime, p deal.StorageDealProposal) {
+
+	// minimum deal duration
+	if p.Duration() < deal.MIN_DEAL_DURATION {
+		rt.Abort("sma._assertValidDealMinimum: deal duration shorter than minimum.")
+	}
+
+	if p.StoragePricePerEpoch() <= deal.MIN_DEAL_PRICE {
+		rt.Abort("sma._assertValidDealMinimum: storage price less than minimum.")
+	}
+
+	// verify StorageDealCollateral match requirements for MinimumStorageDealCollateral
+	if p.ProviderCollateralPerEpoch() < deal.MIN_PROVIDER_DEAL_COLLATERAL_PER_EPOCH ||
+		p.ClientCollateralPerEpoch() < deal.MIN_CLIENT_DEAL_COLLATERAL_PER_EPOCH {
+		rt.Abort("sma._assertValidDealMinimum: deal collaterals less than minimum.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertSufficientBalanceAvailForDeal(rt Runtime, p deal.StorageDealProposal) {
+
+	// verify client and provider has sufficient balance
+	isClientBalAvailable := st._isBalanceAvailable(p.Client(), p.ClientBalanceRequirement())
+	isProviderBalAvailable := st._isBalanceAvailable(p.Provider(), p.ProviderBalanceRequirement())
+
+	if !isClientBalAvailable || !isProviderBalAvailable {
+		rt.Abort("sma._validateNewStorageDeal: client or provider insufficient balance.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertDealExpireAfterMaxProveCommitWindow(rt Runtime, dealP deal.StorageDealProposal) {
+
+	currEpoch := rt.CurrEpoch()
+	dealExpiration := dealP.EndEpoch()
+
+	if dealExpiration <= (currEpoch + sector.MAX_PROVE_COMMIT_SECTOR_EPOCH) {
+		rt.Abort("sma._assertDealExpireAfterMaxProveCommitWindow: deal might expire before prove commit.")
+	}
+
+}
+
+// Call by PublishStorageDeals and GetInitialUtilization (consider remove this)
+// This is the check before a StorageDeal appears onchain
+// It checks the following:
+//   - verify deal did not expire when it is signed
+//   - verify deal hits the chain before StartEpoch
+//   - verify client and provider address and signature are correct (TODO may not be needed)
+//   - verify StorageDealCollateral match requirements for MinimumStorageDealCollateral
+//   - verify client and provider has sufficient balance
+func (st *StorageMarketActorState_I) _validateNewStorageDeal(rt Runtime, d deal.StorageDeal) bool {
+
+	p := d.Proposal()
+
+	st._assertValidClienSignature(rt, p)
+	st._assertValidDealTimingAtPublish(rt, p)
+	st._assertValidDealMinimum(rt, p)
+	st._assertSufficientBalanceAvailForDeal(rt, p)
+
+	return true
+}
+
+func (st *StorageMarketActorState_I) _activateDeal(rt Runtime, deal deal.OnchainDeal) deal.OnchainDeal {
+
+	dealP := deal.Deal().Proposal()
+	deal.Impl().LastPaymentEpoch_ = dealP.StartEpoch()
+	st.Deals()[deal.ID()] = deal
+
+	return deal
+}
+
+// TODO: consider returning a boolean
```

Comment:
otherwise the abort is fine, but may want to call such methods "safe+" as in "safeLockBalance" so it is immediately obvious that the checks happen within the method in code calling it.